src/Compiler/Checking/CheckExpressions.fs

```diff
@@ -6956,16 +6956,27 @@ and TcObjectExpr (cenv: cenv) env tpenv (objTy, realObjTy, argopt, binds, extraI
 //-------------------------------------------------------------------------
 
 /// Check a constant string expression. It might be a 'printf' format string
-and TcConstStringExpr cenv (overallTy: OverallTy) env m tpenv (s: string) =
-
-    let g = cenv.g
+and TcConstStringExpr cenv (overallTy: OverallTy) env m tpenv (s: string) isInlineLiteral =
+    let rec isFormat g typ =
+        match stripTyEqns g typ with
+        | TType_app (tcref, _, _) -> tyconRefEq g tcref g.format4_tcr || tyconRefEq g tcref g.format_tcr
+        | TType_var (typar, _) ->
+            typar.Constraints
+            |> List.exists (fun c ->
+                match c with
+                | TyparConstraint.CoercesTo (typ, _) -> isFormat g typ
+                | _ -> false)
+        | _ -> false
```

Comment:
`TcFormatStringExpr` below creates a type instantiation of the format and does a similar check. I thought this would be faster. We could also move that check to this function, but then a bunch of additional values would have to be passed down to `TcFormatStringExpr`.